core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -58,19 +53,15 @@
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.ScheduledThreadPoolExecutor;
-import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
 
 import static java.util.Collections.unmodifiableMap;
-import static org.elasticsearch.common.unit.SizeValue.parseSizeValue;
-import static org.elasticsearch.common.unit.TimeValue.timeValueMinutes;
 
 /**
  *
  */
-public class ThreadPool extends AbstractComponent implements Closeable {
+public class ThreadPool extends AbstractLifecycleComponent<ThreadPool> {
```

Comment:
It's tricky. I want to construct the executors map once, making it effectively final (although it can not be actually final right now since we do not build it in the constructor). Given this desire, we can not build it in the constructor because we need to wait until all plugins have had a chance to register their custom thread pool settings. But the thread pool instance is needed earlier than this happens because it's needed for the `MonitorService` which is in turn needed for the `NodeModule` which must be built before we start processing plugins. The timeline is thus:

  constructor runs < inject into MonitorService < inject into NodeModule < plugins process modules < start thread pool to create executors

I looked at options around this but I didn't like any of them, they were all super hacky and involved registering suppliers that used the injector.